Race Condition in test_umv

This test fails only sometimes:
```
================================================================================================== FAILURES ==================================================================================================
__________________________________________________________________________________________________ test_umv __________________________________________________________________________________________________

slow_motor = SlowMotor(, name=sim_slow)

    @pytest.mark.timeout(5)
    def test_umv(slow_motor):
        logger.debug('test_umv')
        slow_motor._set_position(5)
>       slow_motor.umvr(2)

tests/test_interface.py:41:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pcdsdevices/interface.py:299: in umvr
    self.umv(delta + self.wm(), timeout=timeout)
pcdsdevices/interface.py:280: in umv
    status_wait(status)
/reg/neh/home/zlentz/miniconda3/envs/dev/lib/python3.6/site-packages/ophyd/status.py:894: in wait
    return status.wait(timeout)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = MoveStatus(done=True, pos=sim_slow, elapsed=0.1, success=False, settle_time=0.0), timeout = None

    def wait(self, timeout=None):
        """
        Block until the action completes.

        When the action has finished succesfully, return ``None``. If the
        action has failed, raise the exception.

        Parameters
        ----------
        timeout: Union[Number, None], optional
            If None (default) wait indefinitely until the status finishes.

        Raises
        ------
        WaitTimeoutError
            If the status has not completed within ``timeout`` (starting from
            when this method was called, not from the beginning of the action).
        StatusTimeoutError
            If the status has failed because the *timeout* that it was
            initialized with has expired.
        Exception
            This is ``status.exception()``, raised if the status has finished
            with an error.  This may include ``TimeoutError``, which
            indicates that the action itself raised ``TimeoutError``, distinct
            from ``WaitTimeoutError`` above.
        """
        if not self._event.wait(timeout=timeout):
            raise WaitTimeoutError("Status has not completed yet.")
        if self._exception is not None:
>           raise self._exception
E           ophyd.utils.errors.UnknownStatusFailure: The status MoveStatus(done=False, pos=sim_slow, elapsed=0.1, success=False, settle_time=0.0) has failed. To obtain more specific, helpful errors in the future, update the Device to use set_exception(...) instead of _finished(success=False).
```

--- a/docs/source/upcoming_release_notes/527-tweak-fixes.rst
+++ b/docs/source/upcoming_release_notes/527-tweak-fixes.rst
@@ -1,0 +1,32 @@
+527 Tweak tweaks
+################
+
+API Changes
+-----------
+- N/A
+
+Features
+--------
+- N/A
+
+Device Updates
+--------------
+- N/A
+
+New Devices
+-----------
+- N/A
+
+Bugfixes
+--------
+- Tweak will feel less "janky" now and give useful feedback.
+- Tweak now accepts + and - as valid inputs for changing the step size.
+- Tweak properly clears lines between prints.
+
+Maintenance
+-----------
+- Tweak refactored for maintainability.
+
+Contributors
+------------
+- N/A

--- a/docs/source/upcoming_release_notes/553-umv-fixes.rst
+++ b/docs/source/upcoming_release_notes/553-umv-fixes.rst
@@ -1,0 +1,30 @@
+IssueNumber Title
+#################
+
+API Changes
+-----------
+- N/A
+
+Features
+--------
+- N/A
+
+Device Updates
+--------------
+- N/A
+
+New Devices
+-----------
+- N/A
+
+Bugfixes
+--------
+- umv will now properly display position and completion status after a move.
+
+Maintenance
+-----------
+- N/A
+
+Contributors
+------------
+- N/A

--- a/docs/source/upcoming_release_notes/553-umv-fixes.rst
+++ b/docs/source/upcoming_release_notes/553-umv-fixes.rst
@@ -1,5 +1,5 @@
-IssueNumber Title
-#################
+553 Umv Fixes
+#############
 
 API Changes
 -----------

--- a/pcdsdevices/interface.py
+++ b/pcdsdevices/interface.py
@@ -1173,14 +1173,25 @@ def tweak_base(*args):
     print()
 
 
+loglist = []
+
+
 class AbsProgressBar(ProgressBar):
     """Progress bar that displays the absolute position as well."""
+    def __init__(self, status_objs, delay_draw=1.0):
+        self._last_position = None
+        super().__init__(status_objs, delay_draw=delay_draw)
+
     def update(self, *args, name=None, current=None, **kwargs):
-        if None not in (name, current):
-            super().update(*args, name='{} ({:.3f})'.format(name, current),
-                           current=current, **kwargs)
-        else:
-            super().update(*args, name=name, current=current, **kwargs)
+        try:
+            name = '{} ({:.3f})'.format(name, current)
+            self._last_position = current
+        except Exception:
+            name = name or 'motor'
+
+        current = current or self._last_position
+        loglist.append((name, current, kwargs))
+        super().update(*args, name=name, current=current, **kwargs)
 
 
 class LightpathMixin(OphydObject):

--- a/pcdsdevices/interface.py
+++ b/pcdsdevices/interface.py
@@ -9,7 +9,7 @@ import signal
 import time
 from contextlib import contextmanager
 from pathlib import Path
-from threading import Event, Thread
+from threading import Event
 from types import MethodType, SimpleNamespace
 from weakref import WeakSet
 
@@ -20,7 +20,7 @@ from ophyd.ophydobj import Kind, OphydObject
 from ophyd.signal import AttributeSignal, Signal
 from ophyd.status import wait as status_wait
 
-from . import utils as util
+from . import utils
 from .signal import NotImplementedSignal
 
 try:
@@ -405,7 +405,10 @@ class MvInterface(BaseInterface):
     def _log_move(self, position):
         logger.info('Moving %s from %s to %s', self.name, self.wm(), position)
 
-    def mv(self, position, timeout=None, wait=False):
+    def _log_move_end(self):
+        logger.info('%s reached position %s', self.name, self.wm())
+
+    def mv(self, position, timeout=None, wait=False, log=True):
         """
         Absolute move to a position.
 
@@ -422,15 +425,21 @@ class MvInterface(BaseInterface):
         wait : bool, optional
             If `True`, wait for motion completion before returning.
             Defaults to :keyword:`False`.
+
+        log : bool, optional
+            If `True`, logs the move at INFO level.
         """
-        self._log_move(position)
+        if log:
+            self._log_move(position)
         self.move(position, timeout=timeout, wait=wait)
+        if wait and log:
+            self._log_move_end()
 
     def wm(self):
         """Get the mover's current positon (where motor)."""
         return self.position
 
-    def __call__(self, position=None, timeout=None, wait=False):
+    def __call__(self, position=None, timeout=None, wait=False, log=True):
         """
         Dispatches to :meth:`mv` or :meth:`wm` based on the arguments.
 
@@ -442,7 +451,7 @@ class MvInterface(BaseInterface):
         if position is None:
             return self.wm()
         else:
-            self.mv(position, timeout=timeout, wait=wait)
+            self.mv(position, timeout=timeout, wait=wait, log=log)
 
     def camonitor(self):
         """
@@ -501,7 +510,7 @@ class FltMvInterface(MvInterface):
             self._presets = Presets(self)
         return self._presets
 
-    def mvr(self, delta, timeout=None, wait=False):
+    def mvr(self, delta, timeout=None, wait=False, log=True):
         """
         Relative move from this position.
 
@@ -518,11 +527,14 @@ class FltMvInterface(MvInterface):
         wait : bool, optional
             If `True`, wait for motion completion before returning.
             Defaults to :keyword:`False`.
+
+        log : bool, optional
+            If `True`, logs the move at INFO level.
         """
 
-        self.mv(delta + self.wm(), timeout=timeout, wait=wait)
+        self.mv(delta + self.wm(), timeout=timeout, wait=wait, log=log)
 
-    def umv(self, position, timeout=None):
+    def umv(self, position, timeout=None, log=True, newline=True):
         """
         Move to a position, wait, and update with a progress bar.
 
@@ -535,19 +547,33 @@ class FltMvInterface(MvInterface):
             If provided, the mover will throw an error if motion takes longer
             than timeout to complete. If omitted, the mover's default timeout
             will be use.
+
+        log : bool, optional
+            If True, logs the move at INFO level.
+
+        newline : bool, optional
+            If True, inserts a newline after the updates.
         """
 
-        self._log_move(position)
+        if log:
+            self._log_move(position)
         status = self.move(position, timeout=timeout, wait=False)
         pgb = AbsProgressBar([status])
         try:
             status_wait(status)
-        except KeyboardInterrupt:
-            self.stop()
+            # Avoid race conditions involving the final update
+            pgb.manual_update()
             pgb.no_more_updates()
-        print()
+        except KeyboardInterrupt:
+            pgb.no_more_updates()
+            self.stop()
+        if pgb.has_updated and newline:
+            # If we made progress bar prints, we need an extra newline
+            print()
+        if log:
+            self._log_move_end()
 
-    def umvr(self, delta, timeout=None):
+    def umvr(self, delta, timeout=None, log=True, newline=True):
         """
         Relative move from this position, wait, and update with a progress bar.
 
@@ -560,9 +586,15 @@ class FltMvInterface(MvInterface):
             If provided, the mover will throw an error if motion takes longer
             than timeout to complete. If omitted, the mover's default timeout
             will be use.
+
+        log : bool, optional
+            If True, logs the move at INFO level.
+
+        newline : bool, optional
+            If True, inserts a newline after the updates.
         """
 
-        self.umv(delta + self.wm(), timeout=timeout)
+        self.umv(delta + self.wm(), timeout=timeout, log=log, newline=newline)
 
     def mv_ginput(self, timeout=None):
         """
@@ -1090,49 +1122,84 @@ def tweak_base(*args):
     """
     Base function to control motors with the arrow keys.
 
-    With one motor, this will use the left and right arrows for the axis and up
-    and down arrows for scaling the step size. With two motors, this will use
-    left and right for the first axis and up and down for the second axis, with
-    shift+arrow used for scaling the step size. The q key quits, as does
-    ctrl+c.
+    With one motor, you can use the right and left arrow keys to move + and -.
+    With two motors, you can also use the up and down arrow keys for the second
+    motor.
+    Three motor and more are not yet supported.
+
+    The scale for the tweak can be doubled by pressing + and halved by pressing
+    -. Shift+up and shift+down can also be used, and the up and down keys will
+    also adjust the scaling in one motor mode.
+
+    Ctrl+c will stop an ongoing move during a tweak without exiting the tweak.
+    Both q and ctrl+c will quit the tweak between moves.
     """
 
-    up = util.arrow_up
-    down = util.arrow_down
-    left = util.arrow_left
-    right = util.arrow_right
-    shift_up = util.shift_arrow_up
-    shift_down = util.shift_arrow_down
+    up = utils.arrow_up
+    down = utils.arrow_down
+    left = utils.arrow_left
+    right = utils.arrow_right
+    shift_up = utils.shift_arrow_up
+    shift_down = utils.shift_arrow_down
+    plus = utils.plus
+    minus = utils.minus
     scale = 0.1
+    abs_status = '{}: {:.4f}'
+    exp_status = '{}: {:.4e}'
 
-    def thread_event():
-        """Function call camonitor to display motor position."""
-        thrd = Thread(target=args[0].camonitor,)
-        thrd.start()
-        args[0]._mov_ev.set()
+    if len(args) == 1:
+        move_keys = (left, right)
+        scale_keys = (up, down, plus, minus, shift_up, shift_down)
+    elif len(args) == 2:
+        move_keys = (left, right, up, down)
+        scale_keys = (plus, minus, shift_up, shift_down)
 
-    def _scale(scale, direction):
+    def show_status():
+        if scale >= 0.0001:
+            template = abs_status
+        else:
+            template = exp_status
+        text = [template.format(mot.name, mot.wm()) for mot in args]
+        text.append(f'scale: {scale}')
+        print('\x1b[2K\r' + ', '.join(text), end='')
+
+    def usage():
+        print()  # Newline
+        if len(args) == 1:
+            print(" Left: move x motor backward")
+            print(" Right: move x motor forward")
+            print(" Up or +: scale*2")
+            print(" Down or -: scale/2")
+        else:
+            print(" Left: move x motor left")
+            print(" Right: move x motor right")
+            print(" Down: move y motor down")
+            print(" Up: move y motor up")
+            print(" + or Shift_Up: scale*2")
+            print(" - or Shift_Down: scale/2")
+        print(" Press q to quit."
+              " Press any other key to display this message.")
+        print()  # Newline
+
+    def edit_scale(scale, direction):
         """Function used to change the scale."""
-        if direction == up or direction == shift_up:
+        if direction in (up, shift_up, plus):
             scale = scale*2
-            print("\r {0:4f}".format(scale), end=" ")
-        elif direction == down or direction == shift_down:
+        elif direction in (down, shift_down, minus):
             scale = scale/2
-            print("\r {0:4f}".format(scale), end=" ")
         return scale
 
     def movement(scale, direction):
         """Function used to know when and the direction to move the motor."""
         try:
             if direction == left:
-                args[0].umvr(-scale)
-                thread_event()
+                args[0].umvr(-scale, log=False, newline=False)
             elif direction == right:
-                args[0].umvr(scale)
-                thread_event()
-            elif direction == up and len(args) > 1:
-                args[1].umvr(scale)
-                print("\r {0:4f}".format(args[1].position), end=" ")
+                args[0].umvr(scale, log=False, newline=False)
+            elif direction == up:
+                args[1].umvr(scale, log=False, newline=False)
+            elif direction == down:
+                args[1].umvr(-scale, log=False, newline=False)
         except Exception as exc:
             logger.error('Error in tweak move: %s', exc)
             logger.debug('', exc_info=True)
@@ -1144,69 +1211,79 @@ def tweak_base(*args):
         logger.info('Started tweak of %s', [mot.name for mot in args])
     is_input = True
     while is_input is True:
-        inp = util.get_input()
+        show_status()
+        inp = utils.get_input()
         if inp in ('q', None):
             is_input = False
         else:
-            if len(args) > 1 and inp == down:
-                movement(-scale, up)
-            elif len(args) > 1 and inp == up:
+            if inp in move_keys:
                 movement(scale, inp)
-            elif inp not in (up, down, left, right, shift_down, shift_up):
-                print()  # Newline
-                if len(args) == 1:
-                    print(" Left: move x motor backward")
-                    print(" Right: move x motor forward")
-                    print(" Up: scale*2")
-                    print(" Down: scale/2")
-                else:
-                    print(" Left: move x motor left")
-                    print(" Right: move x motor right")
-                    print(" Down: move y motor down")
-                    print(" Up: move y motor up")
-                    print(" Shift_Up: scale*2")
-                    print(" Shift_Down: scale/2")
-                print(" Press q to quit."
-                      " Press any other key to display this message.")
-                print()  # Newline
+            elif inp in scale_keys:
+                scale = edit_scale(scale, inp)
             else:
-                movement(scale, inp)
-                scale = _scale(scale, inp)
+                usage()
     print()
 
 
 class AbsProgressBar(ProgressBar):
     """Progress bar that displays the absolute position as well."""
-    def __init__(self, status_objs, **kwargs):
+    def __init__(self, *args, **kwargs):
         self._last_position = None
         self._name = None
         self._no_more = False
-        super().__init__(status_objs, **kwargs)
+        self._manual_cbs = []
+        self.has_updated = False
+        super().__init__(*args, **kwargs)
 
-        # Extra call when status is complete
-        for i, obj in enumerate(status_objs):
-            if not obj.done:
-                obj.add_callback(functools.partial(self._status_cb, i))
+        # Allow manual updates for a final status print
+        for i, obj in enumerate(self.status_objs):
+            self._manual_cbs.append(functools.partial(self._status_cb, i))
 
     def _status_cb(self, pos, status):
         self.update(pos, name=self._name, current=self._last_position)
 
     def update(self, *args, name=None, current=None, **kwargs):
+        # Escape hatch to avoid post-command prints
         if self._no_more:
             return
+
+        # Get cached position and name so they can always be displayed
         current = current or self._last_position
         self._name = self._name or name
+        name = self._name
 
         try:
-            name = '{} ({:.3f})'.format(name, current)
+            # Expand name to include position to display with progress bar
+            name = '{}: {:.4f}'.format(name, current)
             self._last_position = current
         except Exception:
+            # Fallback if there is no position data at all
             name = name or self._name or 'motor'
 
-        super().update(*args, name=name, current=current, **kwargs)
+        try:
+            # Actually draw the bar
+            super().update(*args, name=name, current=current, **kwargs)
+            if not self._no_more:
+                self.has_updated = True
+        except Exception:
+            # Print method failure should never print junk to the screen
+            logger.debug('Error in progress bar update', exc_info=True)
+
+    def manual_update(self):
+        """Execute a manual update of the progress bar."""
+        for cb, status in zip(self._manual_cbs, self.status_objs):
+            cb(status)
 
     def no_more_updates(self):
+        """Prevent all future prints from the progress bar."""
+        self.fp = NullFile()
         self._no_more = True
+        self._manual_cbs.clear()
+
+
+class NullFile:
+    def write(*args, **kwargs):
+        pass
 
 
 class LightpathMixin(OphydObject):
@@ -1268,8 +1345,8 @@ class LightpathMixin(OphydObject):
                     # Use this when the device wasn't ready to set states
                     kw = dict(obj=obj)
                     kw.update(kwargs)
-                    util.schedule_task(self._update_lightpath,
-                                       args=args, kwargs=kw, delay=0.2)
+                    utils.schedule_task(self._update_lightpath,
+                                        args=args, kwargs=kw, delay=0.2)
         except Exception:
             # Without this, callbacks fail silently
             logger.exception('Error in lightpath update callback for %s.',

--- a/pcdsdevices/interface.py
+++ b/pcdsdevices/interface.py
@@ -1204,25 +1204,24 @@ def tweak_base(*args):
             logger.error('Error in tweak move: %s', exc)
             logger.debug('', exc_info=True)
 
+    start_text = ['{} at {:.4f}'.format(mot.name, mot.wm()) for mot in args]
+    logger.info('Started tweak of ' + ', '.join(start_text))
+
     # Loop takes in user key input and stops when 'q' is pressed
-    if len(args) == 1:
-        logger.info('Started tweak of %s', args[0])
-    else:
-        logger.info('Started tweak of %s', [mot.name for mot in args])
     is_input = True
     while is_input is True:
         show_status()
         inp = utils.get_input()
         if inp in ('q', None):
             is_input = False
+        elif inp in move_keys:
+            movement(scale, inp)
+        elif inp in scale_keys:
+            scale = edit_scale(scale, inp)
         else:
-            if inp in move_keys:
-                movement(scale, inp)
-            elif inp in scale_keys:
-                scale = edit_scale(scale, inp)
-            else:
-                usage()
+            usage()
     print()
+    logger.info('Tweak complete')
 
 
 class AbsProgressBar(ProgressBar):

--- a/pcdsdevices/interface.py
+++ b/pcdsdevices/interface.py
@@ -539,11 +539,12 @@ class FltMvInterface(MvInterface):
 
         self._log_move(position)
         status = self.move(position, timeout=timeout, wait=False)
-        AbsProgressBar([status])
+        pgb = AbsProgressBar([status])
         try:
             status_wait(status)
         except KeyboardInterrupt:
             self.stop()
+            pgb.no_more_updates()
         print()
 
     def umvr(self, delta, timeout=None):
@@ -1179,6 +1180,7 @@ class AbsProgressBar(ProgressBar):
     def __init__(self, status_objs, **kwargs):
         self._last_position = None
         self._name = None
+        self._no_more = False
         super().__init__(status_objs, **kwargs)
 
         # Extra call when status is complete
@@ -1189,6 +1191,8 @@ class AbsProgressBar(ProgressBar):
         self.update(pos, name=self._name, current=self._last_position)
 
     def update(self, *args, name=None, current=None, **kwargs):
+        if self._no_more:
+            return
         current = current or self._last_position
         self._name = self._name or name
 
@@ -1199,6 +1203,9 @@ class AbsProgressBar(ProgressBar):
             name = name or self._name or 'motor'
 
         super().update(*args, name=name, current=current, **kwargs)
+
+    def no_more_updates(self):
+        self._no_more = True
 
 
 class LightpathMixin(OphydObject):

--- a/pcdsdevices/interface.py
+++ b/pcdsdevices/interface.py
@@ -1173,24 +1173,30 @@ def tweak_base(*args):
     print()
 
 
-loglist = []
-
-
 class AbsProgressBar(ProgressBar):
     """Progress bar that displays the absolute position as well."""
-    def __init__(self, status_objs, delay_draw=1.0):
+    def __init__(self, status_objs, **kwargs):
         self._last_position = None
-        super().__init__(status_objs, delay_draw=delay_draw)
+        self._name = None
+        super().__init__(status_objs, **kwargs)
+
+        # Extra call when status is complete
+        for i, obj in enumerate(status_objs):
+            obj.add_callback(functools.partial(self._status_cb, i))
+
+    def _status_cb(self, pos, status):
+        self.update(pos, name=self._name, current=self._last_position)
 
     def update(self, *args, name=None, current=None, **kwargs):
+        current = current or self._last_position
+        self._name = self._name or name
+
         try:
             name = '{} ({:.3f})'.format(name, current)
             self._last_position = current
         except Exception:
-            name = name or 'motor'
+            name = name or self._name or 'motor'
 
-        current = current or self._last_position
-        loglist.append((name, current, kwargs))
         super().update(*args, name=name, current=current, **kwargs)
 
 

--- a/pcdsdevices/interface.py
+++ b/pcdsdevices/interface.py
@@ -544,6 +544,7 @@ class FltMvInterface(MvInterface):
             status_wait(status)
         except KeyboardInterrupt:
             self.stop()
+        print()
 
     def umvr(self, delta, timeout=None):
         """

--- a/pcdsdevices/interface.py
+++ b/pcdsdevices/interface.py
@@ -1185,7 +1185,8 @@ class AbsProgressBar(ProgressBar):
 
         # Extra call when status is complete
         for i, obj in enumerate(status_objs):
-            obj.add_callback(functools.partial(self._status_cb, i))
+            if not obj.done:
+                obj.add_callback(functools.partial(self._status_cb, i))
 
     def _status_cb(self, pos, status):
         self.update(pos, name=self._name, current=self._last_position)

--- a/pcdsdevices/sim.py
+++ b/pcdsdevices/sim.py
@@ -47,7 +47,10 @@ class SlowMotor(FastMotor):
 
     def _setup_move(self, position, status):
         if self.position is None:
-            return self._set_position(position)
+            # Initialize position during __init__'s set call
+            self._set_position(position)
+            self._done_moving(success=True)
+            return
 
         def update_thread(positioner, goal):
             positioner._moving = True
@@ -58,7 +61,7 @@ class SlowMotor(FastMotor):
                     positioner._set_position(positioner.position - 1)
                 else:
                     positioner._set_position(goal)
-                    positioner._done_moving()
+                    positioner._done_moving(success=True)
                     return
                 time.sleep(0.1)
             positioner._done_moving(success=False)

--- a/pcdsdevices/utils.py
+++ b/pcdsdevices/utils.py
@@ -32,6 +32,8 @@ ctrl_arrow_up = '\x1b[1;5A'
 ctrl_arrow_down = '\x1b[1;5B'
 ctrl_arrow_right = '\x1b[1;5C'
 ctrl_arrow_left = '\x1b[1;5D'
+plus = '+'
+minus = '-'
 
 
 def is_input():

--- a/tests/test_interface.py
+++ b/tests/test_interface.py
@@ -39,9 +39,10 @@ def test_mv(fast_motor):
 @pytest.mark.timeout(5)
 def test_umv(slow_motor):
     logger.debug('test_umv')
-    slow_motor._set_position(5)
-    slow_motor.umvr(1)
-    assert slow_motor.position == 6
+    start_position = slow_motor.position
+    delta = 2
+    slow_motor.umvr(delta)
+    assert slow_motor.position == start_position + delta
 
 
 def test_camonitor(fast_motor):

